exchanges/bittrex/bittrex_websocket.go

```diff
@@ -0,0 +1,627 @@
+
+	var channels []string
+	for i := range channelsToSubscribe {
+		channels = append(channels, channelsToSubscribe[i].Channel)
+	}
+	arguments := make([][]string, 0)
+	arguments = append(arguments, channels)
+	request.Arguments = arguments
+
+	requestString, err := json.Marshal(request)
+	if err != nil {
+		return err
+	}
+	if b.Verbose {
+		log.Debugf(log.ExchangeSys, "%s Sending JSON message - %s\n", b.Name, requestString)
+	}
+	err = b.Websocket.Conn.SendJSONMessage(request)
+	if err != nil {
+		return err
+	}
+
+	b.WsPendingRequests[request.InvocationID] = WsPendingRequest{
+		request,
+		&channelsToSubscribe,
+	}
+
+	return nil
+}
+
+// Unsubscribe sends a websocket message to stop receiving data from the channel
+func (b *Bittrex) Unsubscribe(channelsToUnsubscribe []stream.ChannelSubscription) error {
+	invocationIDCounter++
+	request := WsEventRequest{
+		Hub:          "c3",
+		Method:       "unsubscribe",
+		InvocationID: invocationIDCounter,
+	}
+
+	var channels []string
+	for i := range channelsToUnsubscribe {
+		channels = append(channels, channelsToUnsubscribe[i].Channel)
+	}
+	arguments := make([][]string, 0)
+	arguments = append(arguments, channels)
+	request.Arguments = arguments
+
+	requestString, err := json.Marshal(request)
+	if err != nil {
+		return err
+	}
+	if b.Verbose {
+		log.Debugf(log.ExchangeSys, "%s Sending JSON message - %s\n", b.Name, requestString)
+	}
+	err = b.Websocket.Conn.SendJSONMessage(request)
+	if err != nil {
+		return err
+	}
+
+	b.Websocket.RemoveSuccessfulUnsubscriptions(channelsToUnsubscribe...)
+	return nil
+}
+
+// wsReadData gets and passes on websocket messages for processing
+func (b *Bittrex) wsReadData() {
+	b.Websocket.Wg.Add(1)
+	defer b.Websocket.Wg.Done()
+
+	for {
+		select {
+		case <-b.Websocket.ShutdownC:
+			return
+		default:
+			resp := b.Websocket.Conn.ReadMessage()
+			if resp.Raw == nil {
+				log.Warnf(log.ExchangeSys, "%s Received empty message\n", b.Name)
+				return
+			}
+
+			err := b.wsHandleData(resp.Raw)
+			if err != nil {
+				b.Websocket.DataHandler <- err
+			}
+		}
+	}
+}
+
+func (b *Bittrex) wsDecodeMessage(encodedMessage string, v interface{}) error {
+	raw, err := crypto.Base64Decode(encodedMessage)
+	if err != nil {
+		return err
+	}
+	reader := flate.NewReader(bytes.NewBuffer(raw))
+	message, err := ioutil.ReadAll(reader)
+	if err != nil {
+		return err
+	}
+	err = json.Unmarshal(message, v)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (b *Bittrex) wsHandleResponseData(request WsPendingRequest, respRaw []byte) error {
+	switch request.Method {
+	case "Authenticate":
+		var response WsAuthResponse
+		err := json.Unmarshal(respRaw, &response)
+		if err != nil {
+			log.Warnf(log.ExchangeSys, "%s - wsHandleResponseData - Cannot unmarshal into WsAuthResponse (%s)\n", b.Name, string(respRaw))
+			return err
+		}
+		if !response.R.Success {
+			log.Warnf(log.ExchangeSys, "%s - Unable to authenticate (%s)", b.Name, response.R.ErrorCode)
+			return nil
```

Comment:
remove

Reply:
OK.
I see that I'm not handling the issue that at that point the client is left in an unauthenticated state. What is the proper way to deal with that?

Reply:
Oh sorry, what I meant was remove return nil as it's done on [line 352](https://github.com/thrasher-corp/gocryptotrader/pull/646/files/ae33262fe2da35456b0794015711b082225bc061#diff-b73870b6d51437919a72391b6c4a36c65e86ec0062a4230e5dafe0753a5473cbR352) what you have done here is fine. Maybe add in `b.Websocket.SetCanUseAuthenticatedEndpoints(false)`